In [2]:
!pip install scikit-surprise

  Using cached scikit_surprise-1.1.4.tar.gz (154 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Failed to build scikit-surprise


  error: subprocess-exited-with-error
  
  × Building wheel for scikit-surprise (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [117 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-312
      creating build\lib.win-amd64-cpython-312\surprise
      copying surprise\accuracy.py -> build\lib.win-amd64-cpython-312\surprise
      copying surprise\builtin_datasets.py -> build\lib.win-amd64-cpython-312\surprise
      copying surprise\dataset.py -> build\lib.win-amd64-cpython-312\surprise
      copying surprise\dump.py -> build\lib.win-amd64-cpython-312\surprise
      copying surprise\reader.py -> build\lib.win-amd64-cpython-312\surprise
      copying surprise\trainset.py -> build\lib.win-amd64-cpython-312\surprise
      copying surprise\utils.py -> build\lib.win-amd64-cpython-312\surprise
      copying surprise\__init__.py -> build\lib.win-amd64-cpython-312\surprise
      cop

In [4]:
# Importing the libraries
%matplotlib inline

# For Basic Data Manipulation
import pandas as pd
import numpy as np

# For Data Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# For Model Building
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate

import warnings
warnings.simplefilter('ignore')

ModuleNotFoundError: No module named 'surprise'

In [ ]:
# Reading the dataset
movies = pd.read_csv('/content/drive/MyDrive/data/movies_metadata.csv')
movies.sample(5)

adult belongs_to_collection budget  \
30867  False                   NaN      0   
44272  False                   NaN      0   
29153  False                   NaN      0   
22804  False                   NaN      0   
35292  False                   NaN      0   

                                                  genres  \
30867                                                 []   
44272                      [{'id': 18, 'name': 'Drama'}]   
29153  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
22804  [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
35292  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   

                               homepage      id    imdb_id original_language  \
30867                               NaN   71776  tt0138324                it   
44272                               NaN  392788  tt4699444                fr   
29153  http://www.studlifethemovie.com/  177354  tt1758770                en   
22804                               NaN   31295  tt0088133                en   
35292                               NaN   67927  tt0038497                sv   

                 original_title  \
30867  Bisturi, la mafia bianca   
44272              Voir du pays   
29153                 Stud Life   
22804               Slagskämpen   
35292   Driver dagg faller regn   

                                                overview  ... release_date  \
30867  This Italian film is a searing indictment of t...  ...   1973-08-25   
44272  At the end of their tour of duty in Afghanista...  ...   2016-09-07   
29153  JJ is a 'Stud' Lesbian. Together with her best...  ...   2012-03-22   
22804  In 1981, after the successful test of a submar...  ...   1985-01-01   
35292  18th century in the north of Sweden. A story o...  ...   1946-12-26   

      revenue runtime                                   spoken_languages  \
30867     0.0     0.0                                                 []   
44272     0.0   102.0  [{'iso_639_1': 'fr', 'name': 'Français'}, {'is...   
29153     0.0    91.0           [{'iso_639_1': 'en', 'name': 'English'}]   
22804     0.0     0.0           [{'iso_639_1': 'en', 'name': 'English'}]   
35292     0.0   102.0           [{'iso_639_1': 'sv', 'name': 'svenska'}]   

         status  tagline                      title  video vote_average  \
30867  Released      NaN  Hospitals:The White Mafia  False          5.0   
44272  Released      NaN               The Stopover  False          6.0   
29153  Released      NaN                  Stud Life  False          5.0   
22804  Released      NaN             The Inside Man  False          5.3   
35292  Released      NaN      Sunshine Follows Rain  False          4.0   

      vote_count  
30867        1.0  
44272        6.0  
29153        1.0  
22804        2.0  
35292        2.0  

[5 rows x 24 columns]

In [ ]:
# Fixing The Genres Column
movies['genres']

0        [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
1        [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...
2        [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...
3        [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...
4                           [{'id': 35, 'name': 'Comedy'}]
                               ...                        
45461    [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...
45462                        [{'id': 18, 'name': 'Drama'}]
45463    [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...
45464                                                   []
45465                                                   []
Name: genres, Length: 45466, dtype: object

In [ ]:
movies['genres'] = movies['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisation of homosexual...  ...   2017-06-09   

           revenue runtime                                   spoken_languages  \
0      373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1      262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2              0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}

I will use IMDB's weighted rating formula to construct my chart. Mathematically, it is represented as follows:

Weighted Rating (WR) =  $\big( \frac{v}{v+m}\cdot R \big) + \big( \frac{v}{v+m}\cdot C \big)$


where,

- v is the number of votes for the movie  
- m is the minimum votes required to be listed in the chart  
- R is the average rating of the movie  
- C is the mean vote across the whole report

To determine an appropriate value for m, the minimum votes required to be listed in the chart. We will use 95th percentile as our cutoff.

i.e. for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

In [ ]:
# Calculating Vote Averages
vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [ ]:
# Calculating minimum votes required
m = vote_counts.quantile(0.95)
m

434.0

In [ ]:
movies['year'] = pd.to_datetime(movies['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)
movies['year']

0        1995
1        1995
2        1995
3        1995
4        1995
         ... 
45461     NaT
45462    2011
45463    2003
45464    1917
45465    2017
Name: year, Length: 45466, dtype: object

In [ ]:
qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified

title  year  vote_count  vote_average  \
0                          Toy Story  1995        5415             7   
1                            Jumanji  1995        2413             6   
5                               Heat  1995        1886             7   
9                          GoldenEye  1995        1194             6   
15                            Casino  1995        1343             7   
...                              ...   ...         ...           ...   
44624        What Happened to Monday  2017         598             7   
44632                  Atomic Blonde  2017         748             6   
44678                        Dunkirk  2017        2712             7   
44842  Transformers: The Last Knight  2017        1440             6   
45014                 The Dark Tower  2017         688             5   

      popularity                                             genres  
0      21.946943                        [Animation, Comedy, Family]  
1      17.015539                       [Adventure, Fantasy, Family]  
5      17.924927                   [Action, Crime, Drama, Thriller]  
9      14.686036                      [Adventure, Action, Thriller]  
15     10.137389                                     [Drama, Crime]  
...          ...                                                ...  
44624  60.581223                        [Science Fiction, Thriller]  
44632  14.455104                                 [Action, Thriller]  
44678  30.938854            [Action, Drama, History, Thriller, War]  
44842  39.186819     [Action, Science Fiction, Thriller, Adventure]  
45014  50.903593  [Action, Western, Science Fiction, Fantasy, Ho...  

[2274 rows x 6 columns]

- To qualify to be considered for the chart, a movie has to have at least 434 votes on TMDB.
- We also see that the average rating for a movie on TMDB is 5.244 on a scale of 10.
- 2274 Movies qualify to be on our chart.

In [ ]:
# Function for calculating weighted rating
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [ ]:
qualified['weighted_rating'] = qualified.apply(weighted_rating, axis=1)
qualified

title  year  vote_count  vote_average  \
0                          Toy Story  1995        5415             7   
1                            Jumanji  1995        2413             6   
5                               Heat  1995        1886             7   
9                          GoldenEye  1995        1194             6   
15                            Casino  1995        1343             7   
...                              ...   ...         ...           ...   
44624        What Happened to Monday  2017         598             7   
44632                  Atomic Blonde  2017         748             6   
44678                        Dunkirk  2017        2712             7   
44842  Transformers: The Last Knight  2017        1440             6   
45014                 The Dark Tower  2017         688             5   

      popularity                                             genres  \
0      21.946943                        [Animation, Comedy, Family]   
1      17.015539                       [Adventure, Fantasy, Family]   
5      17.924927                   [Action, Crime, Drama, Thriller]   
9      14.686036                      [Adventure, Action, Thriller]   
15     10.137389                                     [Drama, Crime]   
...          ...                                                ...   
44624  60.581223                        [Science Fiction, Thriller]   
44632  14.455104                                 [Action, Thriller]   
44678  30.938854            [Action, Drama, History, Thriller, War]   
44842  39.186819     [Action, Science Fiction, Thriller, Adventure]   
45014  50.903593  [Action, Western, Science Fiction, Fantasy, Ho...   

       weighted_rating  
0             6.869770  
1             5.884891  
5             6.671675  
9             5.798701  
15            6.571348  
...                ...  
44624         6.261904  
44632         5.722745  
44678         6.757878  
44842         5.825125  
45014         5.094728  

[2274 rows x 7 columns]

In [ ]:
qualified.sort_values('weighted_rating', ascending=False).head(10)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
292                                         Pulp Fiction  1994        8670   
314                             The Shawshank Redemption  1994        8358   
7000       The Lord of the Rings: The Return of the King  2003        8226   
351                                         Forrest Gump  1994        8147   
5814               The Lord of the Rings: The Two Towers  2002        7641   

       vote_average  popularity  \
15480             8   29.108149   
12481             8  123.167259   
22879             8   32.213481   
2843              8   63.869599   
4863              8   32.070725   
292               8  140.950236   
314               8   51.645403   
7000              8   29.324358   
351               8   48.307194   
5814              8   29.423537   

                                                  genres  weighted_rating  
15480  [Action, Thriller, Science Fiction, Mystery, A...         7.917588  
12481                   [Drama, Action, Crime, Thriller]         7.905871  
22879                [Adventure, Drama, Science Fiction]         7.897107  
2843                                             [Drama]         7.881753  
4863                        [Adventure, Fantasy, Action]         7.871787  
292                                    [Thriller, Crime]         7.868660  
314                                       [Drama, Crime]         7.864000  
7000                        [Adventure, Fantasy, Action]         7.861927  
351                             [Comedy, Drama, Romance]         7.860656  
5814                        [Adventure, Fantasy, Action]         7.851924

- We see that three **Christopher Nolan** Films, **Inception, The Dark Knight and Interstellar** occur at the very top of our chart. The chart also indicates a **strong bias of TMDB Users towards particular genres and directors**.

- Let's now construct our function that builds charts for particular genres. For this, we will use relax our default conditions to the 85th percentile instead of 95.

In [ ]:
series = movies.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
series.name = 'genre'
gen_md = movies.drop('genres', axis=1).join(series)
gen_md

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
1      False                                                NaN  65000000   
...      ...                                                ...       ...   
45463  False                                                NaN         0   
45463  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
0      http://toystory.disney.com/toy-story     862  tt0114709   
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
1                                       NaN    8844  tt0113497   
...                                     ...     ...        ...   
45463                                   NaN   67758  tt0303758   
45463                                   NaN   67758  tt0303758   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language       original_title  \
0                    en            Toy Story   
0                    en            Toy Story   
0                    en            Toy Story   
1                    en              Jumanji   
1                    en              Jumanji   
...                 ...                  ...   
45463                en             Betrayal   
45463                en             Betrayal   
45463                en             Betrayal   
45464                en  Satana likuyushchiy   
45465                en             Queerama   

                                                overview popularity  ...  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
0      Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
0      Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
1      When siblings Judy and Peter discover an encha...  17.015539  ...   
1      When siblings Judy and Peter discover an encha...  17.015539  ...   
...                                                  ...        ...  ...   
45463  When one of her hits goes wrong, a professiona...   0.903007  ...   
45463  When one of her hits goes wrong, a professiona...   0.903007  ...   
45463  When one of her hits goes wrong, a professiona...   0.903007  ...   
45464  In a small town live two brothers, one a minis...   0.003503  ...   
45465  50 years after decriminalisation of homosexual...   0.163015  ...   

      runtime                                   spoken_languages    status  \
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
...       ...                                                ...       ...   
45463    90.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
45463    90.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
45463    90.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
45464    87.0      

In [ ]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)

    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')

    qualified['weighted_rating'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('weighted_rating', ascending=False).head(250)

    return qualified

**Top Romance Movies**

In [ ]:
build_chart('Romance').head(10)

title  year  vote_count  vote_average popularity  \
10309  Dilwale Dulhania Le Jayenge  1995         661             9  34.457024   
351                   Forrest Gump  1994        8147             8  48.307194   
876                        Vertigo  1958        1162             8   18.20822   
40251                   Your Name.  2016        1030             8  34.461252   
883               Some Like It Hot  1959         835             8  11.845107   
1132               Cinema Paradiso  1988         834             8  14.177005   
19901                     Paperman  2012         734             8   7.198633   
37863                  Sing Street  2016         669             8  10.672862   
882                  The Apartment  1960         498             8  11.994281   
38718               The Handmaiden  2016         453             8  16.727405   

       weighted_rating  
10309         8.565285  
351           7.971357  
876           7.811667  
40251         7.789489  
883           7.745154  
1132          7.744878  
19901         7.713951  
37863         7.689483  
882           7.599317  
38718         7.566166

In [ ]:
build_chart('Action').head(10)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
7000       The Lord of the Rings: The Return of the King  2003        8226   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1154                             The Empire Strikes Back  1980        5998   
4135                                            Scarface  1983        3017   
9430                                              Oldboy  2003        2000   
1910                                       Seven Samurai  1954         892   

       vote_average  popularity  weighted_rating  
15480             8   29.108149         7.955099  
12481             8  123.167259         7.948610  
4863              8   32.070725         7.929579  
7000              8   29.324358         7.924031  
5814              8   29.423537         7.918382  
256               8   42.149697         7.908327  
1154              8   19.470959         7.896841  
4135              8   11.299673         7.802046  
9430              8   10.616859         7.711649  
1910              8    15.01777         7.426145

## Content Based Recommender

The recommender we built in the previous section suffers some severe limitations. For one, it gives the same recommendation to everyone, regardless of the user's personal taste. If a person who loves romantic movies (and hates action) were to look at our Top 15 Chart, s/he wouldn't probably like most of the movies. If s/he were to go one step further and look at our charts by genre, s/he wouldn't still be getting the best recommendations.

To personalise our recommendations more, I am going to build an engine that computes similarity between movies based on certain metrics and suggests movies that are most similar to a particular movie that a user liked. Since we will be using movie metadata (or content) to build this engine, this also known as Content Based Filtering.

I will build two Content Based Recommenders based on:

- Movie Overviews and Taglines
- Movie Cast, Crew, Keywords and Genre  

Also, as mentioned in the introduction, I will be using a subset of all the movies available to us due to limiting computing power available to me.

In [ ]:
links_small = pd.read_csv('/content/drive/MyDrive/data/links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')
links_small

0          862
1         8844
2        15602
3        31357
4        11862
         ...  
9120    402672
9121    315011
9122    391698
9123    137608
9124    410803
Name: tmdbId, Length: 9112, dtype: int64

In [ ]:
movies = movies.drop([19730, 29503, 35587])
movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ...      revenue  \
0      Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1      When siblings Judy and Peter discover an encha...  ...  262797249.0   
2      A family wedding reignites the ancient feud be...  ...          0.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4      Just when George Banks has recovered from his ...  ...   76578911.0   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          0.0   
45462  An artist struggles to finish his work while a...  ...          0.0   
45463  When one of her hits goes wrong, a professiona...  ...          0.0   
45464  In a small town live two brothers, one a minis...  ...          0.0   
45465  50 years after decriminalisation of homosexual...  ...          0.0   

      runtime                                   spoken_languages    status  \
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2       101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3      

In [ ]:
movies['id'] = movies['id'].astype('int')
movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ...      revenue  \
0      Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1      When siblings Judy and Peter discover an encha...  ...  262797249.0   
2      A family wedding reignites the ancient feud be...  ...          0.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4      Just when George Banks has recovered from his ...  ...   76578911.0   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          0.0   
45462  An artist struggles to finish his work while a...  ...          0.0   
45463  When one of her hits goes wrong, a professiona...  ...          0.0   
45464  In a small town live two brothers, one a minis...  ...          0.0   
45465  50 years after decriminalisation of homosexual...  ...          0.0   

      runtime                                   spoken_languages    status  \
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2       101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3      

In [ ]:
smd = movies[movies['id'].isin(links_small)]
smd

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
40224  False                                                NaN  15000000   
40503  False                                                NaN         0   
44821  False  {'id': 34055, 'name': 'Pokémon Collection', 'p...  16000000   
44826  False  {'id': 34055, 'name': 'Pokémon Collection', 'p...         0   
45265  False                                                NaN         0   

                                                  genres  \
0                            [Animation, Comedy, Family]   
1                           [Adventure, Fantasy, Family]   
2                                      [Romance, Comedy]   
3                               [Comedy, Drama, Romance]   
4                                               [Comedy]   
...                                                  ...   
40224  [Action, Adventure, Drama, Horror, Science Fic...   
40503                               [Documentary, Music]   
44821    [Adventure, Fantasy, Animation, Action, Family]   
44826  [Adventure, Fantasy, Animation, Science Fictio...   
45265                                    [Comedy, Drama]   

                                                homepage      id    imdb_id  \
0                   http://toystory.disney.com/toy-story     862  tt0114709   
1                                                    NaN    8844  tt0113497   
2                                                    NaN   15602  tt0113228   
3                                                    NaN   31357  tt0114885   
4                                                    NaN   11862  tt0113041   
...                                                  ...     ...        ...   
40224                                                NaN  315011  tt4262980   
40503              http://www.thebeatlesliveproject.com/  391698  tt2531318   
44821                  http://movies.warnerbros.com/pk3/   10991  tt0235679   
44826  http://www.pokemon.com/us/movies/movie-pokemon...   12600  tt0287635   
45265                                                NaN  265189  tt2121382   

      original_language                                     original_title  \
0                    en                                          Toy Story   
1                    en                                            Jumanji   
2                    en                                   Grumpier Old Men   
3                    en                                  Waiting to Exhale   
4                    en                        Father of the Bride Part II   
...                 ...                                                ...   
40224                ja                                             シン・ゴジラ   
40503                en  The Beatles: Eight Days a Week - The Touring Y...   
44821                ja                               Pokémon 3: The Movie   
44826                ja                     劇場版ポケットモンスター セレビィ 時を越えた遭遇（であい）   
45265                sv                                             Turist   

                                                overview  ...      revenue  \
0      Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1      When siblings Judy and Peter discover an encha...  ...  262797249.0   
2      A family wedding reignites the ancient feud be...  ...          0.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4      Just when George Banks has recovered from his ...  ...   76578911.0   
...    

# Movie Description Based Recommender

We have 9099 movies avaiable in our small movies metadata dataset which is 5 times smaller than our original dataset of 45000 movies.

In [ ]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [ ]:
smd

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
40224  False                                                NaN  15000000   
40503  False                                                NaN         0   
44821  False  {'id': 34055, 'name': 'Pokémon Collection', 'p...  16000000   
44826  False  {'id': 34055, 'name': 'Pokémon Collection', 'p...         0   
45265  False                                                NaN         0   

                                                  genres  \
0                            [Animation, Comedy, Family]   
1                           [Adventure, Fantasy, Family]   
2                                      [Romance, Comedy]   
3                               [Comedy, Drama, Romance]   
4                                               [Comedy]   
...                                                  ...   
40224  [Action, Adventure, Drama, Horror, Science Fic...   
40503                               [Documentary, Music]   
44821    [Adventure, Fantasy, Animation, Action, Family]   
44826  [Adventure, Fantasy, Animation, Science Fictio...   
45265                                    [Comedy, Drama]   

                                                homepage      id    imdb_id  \
0                   http://toystory.disney.com/toy-story     862  tt0114709   
1                                                    NaN    8844  tt0113497   
2                                                    NaN   15602  tt0113228   
3                                                    NaN   31357  tt0114885   
4                                                    NaN   11862  tt0113041   
...                                                  ...     ...        ...   
40224                                                NaN  315011  tt4262980   
40503              http://www.thebeatlesliveproject.com/  391698  tt2531318   
44821                  http://movies.warnerbros.com/pk3/   10991  tt0235679   
44826  http://www.pokemon.com/us/movies/movie-pokemon...   12600  tt0287635   
45265                                                NaN  265189  tt2121382   

      original_language                                     original_title  \
0                    en                                          Toy Story   
1                    en                                            Jumanji   
2                    en                                   Grumpier Old Men   
3                    en                                  Waiting to Exhale   
4                    en                        Father of the Bride Part II   
...                 ...                                                ...   
40224                ja                                             シン・ゴジラ   
40503                en  The Beatles: Eight Days a Week - The Touring Y...   
44821                ja                               Pokémon 3: The Movie   
44826                ja                     劇場版ポケットモンスター セレビィ 時を越えた遭遇（であい）   
45265                sv                                             Turist   

                                                overview  ... runtime  \
0      Led by Woody, Andy's toys live happily in his ...  ...    81.0   
1      When siblings Judy and Peter discover an encha...  ...   104.0   
2      A family wedding reignites the ancient feud be...  ...   101.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   127.0   
4      Just when George Banks has recovered from his ...  ...   106.0   
...                                  

In [ ]:
# Using TfIdf Vectorizer
tf = TfidfVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tf.fit_transform(smd['description'])
tfidf_matrix.shape

(9099, 268124)

In [ ]:
# Using cosine si,illarity to determine simillarities between description
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)
cosine_sim

array([[1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
        0.        ],
       [0.00680476, 1.        , 0.01531062, ..., 0.00357057, 0.00762326,
        0.        ],
       [0.        , 0.01531062, 1.        , ..., 0.        , 0.00286535,
        0.00472155],
       ...,
       [0.        , 0.00357057, 0.        , ..., 1.        , 0.07811616,
        0.        ],
       [0.00344913, 0.00762326, 0.00286535, ..., 0.07811616, 1.        ,
        0.        ],
       [0.        , 0.        , 0.00472155, ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])
indices

title
Toy Story                                                0
Jumanji                                                  1
Grumpier Old Men                                         2
Waiting to Exhale                                        3
Father of the Bride Part II                              4
                                                      ... 
Shin Godzilla                                         9094
The Beatles: Eight Days a Week - The Touring Years    9095
Pokémon: Spell of the Unknown                         9096
Pokémon 4Ever: Celebi - Voice of the Forest           9097
Force Majeure                                         9098
Length: 9099, dtype: int64

In [ ]:
# Getting the recommendations based on movies description
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

In [ ]:
get_recommendations('The Dark Knight').head(10)

7931                      The Dark Knight Rises
132                              Batman Forever
1113                             Batman Returns
8227    Batman: The Dark Knight Returns, Part 2
7565                 Batman: Under the Red Hood
524                                      Batman
7901                           Batman: Year One
2579               Batman: Mask of the Phantasm
2696                                        JFK
8165    Batman: The Dark Knight Returns, Part 1
Name: title, dtype: object

In [ ]:
get_recommendations('Toy Story').head(10)

2502               Toy Story 2
7535               Toy Story 3
6193    The 40 Year Old Virgin
2547           Man on the Moon
6627              Factory Girl
4702    What's Up, Tiger Lily?
889      Rebel Without a Cause
6554    For Your Consideration
4988          Rivers and Tides
1599                 Condorman
Name: title, dtype: object

# Metadata Based Recommender

In [ ]:
credits = pd.read_csv('/content/drive/MyDrive/data/credits.csv')
keywords = pd.read_csv('/content/drive/MyDrive/data/keywords.csv')

In [ ]:
credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [ ]:
keywords.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [ ]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
movies['id'] = movies['id'].astype('int')

In [ ]:
movies = movies.merge(credits, on='id')
movies = movies.merge(keywords, on='id')

In [ ]:
smd = movies[movies['id'].isin(links_small)]
smd

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
40952  False                                                NaN   8000000   
41172  False                                                NaN   1000000   
41225  False                                                NaN  15050000   
41391  False                                                NaN  15000000   
41669  False                                                NaN         0   

                                                  genres  \
0                            [Animation, Comedy, Family]   
1                           [Adventure, Fantasy, Family]   
2                                      [Romance, Comedy]   
3                               [Comedy, Drama, Romance]   
4                                               [Comedy]   
...                                                  ...   
40952                                            [Drama]   
41172                                [Thriller, Romance]   
41225               [Adventure, Drama, History, Romance]   
41391  [Action, Adventure, Drama, Horror, Science Fic...   
41669                               [Documentary, Music]   

                                    homepage      id    imdb_id  \
0       http://toystory.disney.com/toy-story     862  tt0114709   
1                                        NaN    8844  tt0113497   
2                                        NaN   15602  tt0113228   
3                                        NaN   31357  tt0114885   
4                                        NaN   11862  tt0113041   
...                                      ...     ...        ...   
40952                                    NaN  159550  tt0255313   
41172                                    NaN  392572  tt5165344   
41225                                    NaN  402672  tt3859980   
41391                                    NaN  315011  tt4262980   
41669  http://www.thebeatlesliveproject.com/  391698  tt2531318   

      original_language                                     original_title  \
0                    en                                          Toy Story   
1                    en                                            Jumanji   
2                    en                                   Grumpier Old Men   
3                    en                                  Waiting to Exhale   
4                    en                        Father of the Bride Part II   
...                 ...                                                ...   
40952                en                     The Last Brickmaker in America   
41172                hi                                             रुस्तम   
41225                hi                                       Mohenjo Daro   
41391                ja                                             シン・ゴジラ   
41669                en  The Beatles: Eight Days a Week - The Touring Y...   

                                                overview  ...    status  \
0      Led by Woody, Andy's toys live happily in his ...  ...  Released   
1      When siblings Judy and Peter discover an encha...  ...  Released   
2      A family wedding reignites the ancient feud be...  ...  Released   
3      Cheated on, mistreated and stepped on, the wom...  ...  Released   
4      Just when George Banks has recovered from his ...  ...  Released   
...                                                  ...  ...       ...   
40952  A man must cope with the loss of his wife and ...  ...  Released   
41172  Rustom Pavri

In [ ]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [ ]:
smd

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
40952  False                                                NaN   8000000   
41172  False                                                NaN   1000000   
41225  False                                                NaN  15050000   
41391  False                                                NaN  15000000   
41669  False                                                NaN         0   

                                                  genres  \
0                            [Animation, Comedy, Family]   
1                           [Adventure, Fantasy, Family]   
2                                      [Romance, Comedy]   
3                               [Comedy, Drama, Romance]   
4                                               [Comedy]   
...                                                  ...   
40952                                            [Drama]   
41172                                [Thriller, Romance]   
41225               [Adventure, Drama, History, Romance]   
41391  [Action, Adventure, Drama, Horror, Science Fic...   
41669                               [Documentary, Music]   

                                    homepage      id    imdb_id  \
0       http://toystory.disney.com/toy-story     862  tt0114709   
1                                        NaN    8844  tt0113497   
2                                        NaN   15602  tt0113228   
3                                        NaN   31357  tt0114885   
4                                        NaN   11862  tt0113041   
...                                      ...     ...        ...   
40952                                    NaN  159550  tt0255313   
41172                                    NaN  392572  tt5165344   
41225                                    NaN  402672  tt3859980   
41391                                    NaN  315011  tt4262980   
41669  http://www.thebeatlesliveproject.com/  391698  tt2531318   

      original_language                                     original_title  \
0                    en                                          Toy Story   
1                    en                                            Jumanji   
2                    en                                   Grumpier Old Men   
3                    en                                  Waiting to Exhale   
4                    en                        Father of the Bride Part II   
...                 ...                                                ...   
40952                en                     The Last Brickmaker in America   
41172                hi                                             रुस्तम   
41225                hi                                       Mohenjo Daro   
41391                ja                                             シン・ゴジラ   
41669                en  The Beatles: Eight Days a Week - The Touring Y...   

                                                overview  ...  \
0      Led by Woody, Andy's toys live happily in his ...  ...   
1      When siblings Judy and Peter discover an encha...  ...   
2      A family wedding reignites the ancient feud be...  ...   
3      Cheated on, mistreated and stepped on, the wom...  ...   
4      Just when George Banks has recovered from his ...  ...   
...                                                  ...  ...   
40952  A man must cope with the loss of his wife and ...  ...   
41172  Rustom Pavri, an honourable officer of the Ind...  ...   
41225  Village lad Sarman is drawn

In [ ]:
def get_director_name(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [ ]:
smd['director'] = smd['crew'].apply(get_director_name)

In [ ]:
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)

In [ ]:
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

Now the idea is to create a metadata dump for every movie which consists of genres, director, main actors and keywords. Then use a Count Vectorizer to create our count matrix as we did in the Description Recommender. The remaining steps are similar to what we did earlier: we calculate the cosine similarities and return movies that are most similar.

Steps to follow :  
- Strip Spaces and Convert to Lowercase from all our features. This way, our engine will not confuse between Tom Cruise and Tom Hanks.
- Mention Director 3 times to give it more weight relative to the entire cast.

In [ ]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])
smd

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
40952  False                                                NaN   8000000   
41172  False                                                NaN   1000000   
41225  False                                                NaN  15050000   
41391  False                                                NaN  15000000   
41669  False                                                NaN         0   

                                                  genres  \
0                            [Animation, Comedy, Family]   
1                           [Adventure, Fantasy, Family]   
2                                      [Romance, Comedy]   
3                               [Comedy, Drama, Romance]   
4                                               [Comedy]   
...                                                  ...   
40952                                            [Drama]   
41172                                [Thriller, Romance]   
41225               [Adventure, Drama, History, Romance]   
41391  [Action, Adventure, Drama, Horror, Science Fic...   
41669                               [Documentary, Music]   

                                    homepage      id    imdb_id  \
0       http://toystory.disney.com/toy-story     862  tt0114709   
1                                        NaN    8844  tt0113497   
2                                        NaN   15602  tt0113228   
3                                        NaN   31357  tt0114885   
4                                        NaN   11862  tt0113041   
...                                      ...     ...        ...   
40952                                    NaN  159550  tt0255313   
41172                                    NaN  392572  tt5165344   
41225                                    NaN  402672  tt3859980   
41391                                    NaN  315011  tt4262980   
41669  http://www.thebeatlesliveproject.com/  391698  tt2531318   

      original_language                                     original_title  \
0                    en                                          Toy Story   
1                    en                                            Jumanji   
2                    en                                   Grumpier Old Men   
3                    en                                  Waiting to Exhale   
4                    en                        Father of the Bride Part II   
...                 ...                                                ...   
40952                en                     The Last Brickmaker in America   
41172                hi                                             रुस्तम   
41225                hi                                       Mohenjo Daro   
41391                ja                                             シン・ゴジラ   
41669                en  The Beatles: Eight Days a Week - The Touring Y...   

                                                overview  ...  video  \
0      Led by Woody, Andy's toys live happily in his ...  ...  False   
1      When siblings Judy and Peter discover an encha...  ...  False   
2      A family wedding reignites the ancient feud be...  ...  False   
3      Cheated on, mistreated and stepped on, the wom...  ...  False   
4      Just when George Banks has recovered from his ...  ...  False   
...                                                  ...  ...    ...   
40952  A man must cope with the loss of his wife and ...  ...  False   
41172  Rustom Pavri, an honourable officer 

In [ ]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [ ]:
s.value_counts().head(10)

keyword
independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
violence                264
love                    222
musical                 219
sex                     219
suspense                212
Name: count, dtype: int64

In [ ]:
# Removing the keywords which occur only once.
s = s.value_counts()
s = s[s > 1]
s.head()

keyword
independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: count, dtype: int64

In [ ]:
# Stemming
stemmer = SnowballStemmer('english')
stemmer.stem('dancing')

'danc'

In [ ]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [ ]:
smd.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   
3      [Comedy, Drama, Romance]                                   NaN  31357   
4                      [Comedy]                                   NaN  11862   

     imdb_id original_language               original_title  \
0  tt0114709                en                    Toy Story   
1  tt0113497                en                      Jumanji   
2  tt0113228                en             Grumpier Old Men   
3  tt0114885                en            Waiting to Exhale   
4  tt0113041                en  Father of the Bride Part II   

                                            overview  ...  video vote_average  \
0  Led by Woody, Andy's toys live happily in his ...  ...  False          7.7   
1  When siblings Judy and Peter discover an encha...  ...  False          6.9   
2  A family wedding reignites the ancient feud be...  ...  False          6.5   
3  Cheated on, mistreated and stepped on, the wom...  ...  False          6.1   
4  Just when George Banks has recovered from his ...  ...  False          5.7   

  vote_count  year                                            cast  \
0     5415.0  1995                [tomhanks, timallen, donrickles]   
1     2413.0  1995     [robinwilliams, jonathanhyde, kirstendunst]   
2       92.0  1995        [waltermatthau, jacklemmon, ann-margret]   
3       34.0  1995  [whitneyhouston, angelabassett, lorettadevine]   
4      173.0  1995         [stevemartin, dianekeaton, martinshort]   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...   

                                            keywords cast_size crew_size  \
0  [jealousy, toy, boy, friendship, friends, riva...        13       106   
1  [board game, disappearance, based on children'...        26        16   
2  [fishing, best friend, duringcreditsstinger, o...         7         4   
3  [based on novel, interracial relationship, sin...        10        10   
4  [baby, midlife crisis, confidence, aging, daug...        12         7   

                                           director  
0        [johnlasseter, johnlasseter, johnlasseter]  
1           [joejohnston, joejohnston, joejohnston]  
2        [howarddeutch, howarddeutch, howarddeutch]  
3  [forestwhitaker, forestwhitaker, forestwhitaker]  
4        [charlesshyer, charlesshyer, charlesshyer]  

[5 rows x 31 columns]

In [ ]:
smd['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [ ]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))
smd

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
40952  False                                                NaN   8000000   
41172  False                                                NaN   1000000   
41225  False                                                NaN  15050000   
41391  False                                                NaN  15000000   
41669  False                                                NaN         0   

                                                  genres  \
0                            [Animation, Comedy, Family]   
1                           [Adventure, Fantasy, Family]   
2                                      [Romance, Comedy]   
3                               [Comedy, Drama, Romance]   
4                                               [Comedy]   
...                                                  ...   
40952                                            [Drama]   
41172                                [Thriller, Romance]   
41225               [Adventure, Drama, History, Romance]   
41391  [Action, Adventure, Drama, Horror, Science Fic...   
41669                               [Documentary, Music]   

                                    homepage      id    imdb_id  \
0       http://toystory.disney.com/toy-story     862  tt0114709   
1                                        NaN    8844  tt0113497   
2                                        NaN   15602  tt0113228   
3                                        NaN   31357  tt0114885   
4                                        NaN   11862  tt0113041   
...                                      ...     ...        ...   
40952                                    NaN  159550  tt0255313   
41172                                    NaN  392572  tt5165344   
41225                                    NaN  402672  tt3859980   
41391                                    NaN  315011  tt4262980   
41669  http://www.thebeatlesliveproject.com/  391698  tt2531318   

      original_language                                     original_title  \
0                    en                                          Toy Story   
1                    en                                            Jumanji   
2                    en                                   Grumpier Old Men   
3                    en                                  Waiting to Exhale   
4                    en                        Father of the Bride Part II   
...                 ...                                                ...   
40952                en                     The Last Brickmaker in America   
41172                hi                                             रुस्तम   
41225                hi                                       Mohenjo Daro   
41391                ja                                             シン・ゴジラ   
41669                en  The Beatles: Eight Days a Week - The Touring Y...   

                                                overview  ... vote_average  \
0      Led by Woody, Andy's toys live happily in his ...  ...          7.7   
1      When siblings Judy and Peter discover an encha...  ...          6.9   
2      A family wedding reignites the ancient feud be...  ...          6.5   
3      Cheated on, mistreated and stepped on, the wom...  ...          6.1   
4      Just when George Banks has recovered from his ...  ...          5.7   
...                                                  ...  ...          ...   
40952  A man must cope with the loss of his wife and ...  ...          7.

In [ ]:
count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [ ]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)
cosine_sim

array([[1.        , 0.02441931, 0.02738955, ..., 0.        , 0.        ,
        0.        ],
       [0.02441931, 1.        , 0.        , ..., 0.02973505, 0.02500782,
        0.        ],
       [0.02738955, 0.        , 1.        , ..., 0.03335187, 0.        ,
        0.        ],
       ...,
       [0.        , 0.02973505, 0.03335187, ..., 1.        , 0.08700222,
        0.        ],
       [0.        , 0.02500782, 0.        , ..., 0.08700222, 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [ ]:
indices

title
Toy Story                                                0
Jumanji                                                  1
Grumpier Old Men                                         2
Waiting to Exhale                                        3
Father of the Bride Part II                              4
                                                      ... 
The Last Brickmaker in America                        9214
Rustom                                                9215
Mohenjo Daro                                          9216
Shin Godzilla                                         9217
The Beatles: Eight Days a Week - The Touring Years    9218
Length: 9219, dtype: int64

In [ ]:
get_recommendations('The Dark Knight').head(10)

8031         The Dark Knight Rises
6218                 Batman Begins
6623                  The Prestige
2085                     Following
7648                     Inception
4145                      Insomnia
3381                       Memento
8613                  Interstellar
7659    Batman: Under the Red Hood
1134                Batman Returns
Name: title, dtype: object

In [ ]:
get_recommendations('Dumb and Dumber').head(10)

8808          Dumb and Dumber To
7187                     Yes Man
132               Batman Forever
8645             The Other Woman
2181                American Pie
3486                  Sexy Beast
1316          A Smile Like Yours
4416    Varning för Jönssonligan
6448                      Tsotsi
4830                 Once Bitten
Name: title, dtype: object

In [ ]:
def improved_recommendations(title):
    idx = indices[title]

    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]

    movie_indices = [i[0] for i in sim_scores]

    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')

    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)

    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['weighted_rating'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('weighted_rating', ascending=False).head(10)

    return qualified

In [ ]:
improved_recommendations('The Dark Knight')

title  vote_count  vote_average  year  \
7648                           Inception       14075             8  2010   
8613                        Interstellar       11187             8  2014   
6623                        The Prestige        4510             8  2006   
3381                             Memento        4168             8  2000   
8031               The Dark Knight Rises        9263             7  2012   
6218                       Batman Begins        7511             7  2005   
1134                      Batman Returns        1706             6  1992   
132                       Batman Forever        1529             5  1995   
9024  Batman v Superman: Dawn of Justice        7189             5  2016   
1260                      Batman & Robin        1447             4  1997   

      weighted_rating  
7648         7.917588  
8613         7.897107  
6623         7.758148  
3381         7.740175  
8031         6.921448  
6218         6.904127  
1134         5.846862  
132          5.054144  
9024         5.013943  
1260         4.287233

In [ ]:
improved_recommendations('Face/Off')

title  vote_count  vote_average  year  weighted_rating
142                Desperado         898             6  1995         5.753968
7571                Repo Men         627             6  2010         5.691126
4161             Windtalkers         345             5  2002         5.136438
3834                 The One         445             5  2001         5.120916
87              Broken Arrow         459             5  1996         5.119020
2413               RoboCop 2         498             5  1990         5.114040
5082                Paycheck         594             5  2003         5.103390
6716                    Next         880             5  2007         5.080887
4537               Daredevil        1551             5  2003         5.053544
2909  Mission: Impossible II        1966             5  2000         5.044285

# Collaborative Filtering

In [ ]:
# Using surprise library
reader = Reader()
ratings = pd.read_csv('/content/drive/MyDrive/data/ratings_small.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [ ]:
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)


In [ ]:
data

In [ ]:
svd = SVD()
cross_validate(svd, data, measures=['RMSE', 'MAE'], cv = 10)

{'test_rmse': array([0.88922177, 0.89230263, 0.88317371, 0.89568029, 0.88930141,
        0.88748941, 0.89306597, 0.90820706, 0.89422414, 0.87064596]),
 'test_mae': array([0.67983213, 0.68426204, 0.6777604 , 0.68951056, 0.68636598,
        0.68641219, 0.68166035, 0.70033339, 0.68790636, 0.67561492]),
 'fit_time': (1.2150959968566895,
  1.1830356121063232,
  1.1643290519714355,
  1.1334707736968994,
  1.1432766914367676,
  1.1302170753479004,
  1.841235876083374,
  1.7585408687591553,
  1.1412010192871094,
  1.1406927108764648),
 'test_time': (0.10145306587219238,
  0.04719042778015137,
  0.18549013137817383,
  0.04394388198852539,
  0.04465460777282715,
  0.04332256317138672,
  0.07905340194702148,
  0.1868443489074707,
  0.059964895248413086,
  0.05382561683654785)}

In [ ]:
trainset = data.build_full_trainset()
svd.fit(trainset)

In [ ]:
svd.predict(1, 302, 3)

Prediction(uid=1, iid=302, r_ui=3, est=2.7167816413514436, details={'was_impossible': False})

# Hybrid Recommender


Now we will try to build a simple hybrid recommender that brings together techniques we have implemented in the content based and collaborative filter based engines. This is how it will work:

- Input: User ID and the Title of a Movie
- Output: Similar movies sorted on the basis of expected ratings by that particular user.


In [ ]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [ ]:
id_map = pd.read_csv('/content/drive/MyDrive/data/links_small.csv')[['movieId', 'tmdbId']]
id_map['tmdbId'] = id_map['tmdbId'].apply(convert_int)
id_map.columns = ['movieId', 'id']
id_map = id_map.merge(smd[['title', 'id']], on='id').set_index('title')
id_map

movieId        id
title                                                                
Toy Story                                                 1     862.0
Jumanji                                                   2    8844.0
Grumpier Old Men                                          3   15602.0
Waiting to Exhale                                         4   31357.0
Father of the Bride Part II                               5   11862.0
...                                                     ...       ...
The Last Brickmaker in America                       161944  159550.0
Rustom                                               162542  392572.0
Mohenjo Daro                                         162672  402672.0
Shin Godzilla                                        163056  315011.0
The Beatles: Eight Days a Week - The Touring Years   163949  391698.0

[9219 rows x 2 columns]

In [ ]:
indices_map = id_map.set_index('id')
indices_map

movieId
id               
862.0           1
8844.0          2
15602.0         3
31357.0         4
11862.0         5
...           ...
159550.0   161944
392572.0   162542
402672.0   162672
315011.0   163056
391698.0   163949

[9219 rows x 1 columns]

In [ ]:
def hybrid_recommender(userId, title):
    idx = indices[title]
    tmdbId = id_map.loc[title]['id']
    #print(idx)
    movie_id = id_map.loc[title]['movieId']

    sim_scores = list(enumerate(cosine_sim[int(idx)]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]

    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year', 'id']]
    movies['est'] = movies['id'].apply(lambda x: svd.predict(userId, indices_map.loc[x]['movieId']).est)
    movies = movies.sort_values('est', ascending=False)
    return movies.head(10)

In [ ]:
hybrid_recommender(1, 'Avatar')

title  vote_count  vote_average  year  \
522           Terminator 2: Judgment Day      4274.0           7.7  1991   
1011                      The Terminator      4208.0           7.4  1984   
8401             Star Trek Into Darkness      4479.0           7.4  2013   
974                               Aliens      3282.0           7.7  1986   
8658          X-Men: Days of Future Past      6155.0           7.5  2014   
1376                             Titanic      7770.0           7.5  1997   
922                            The Abyss       822.0           7.1  1989   
2014                    Fantastic Planet       140.0           7.6  1973   
1621  Darby O'Gill and the Little People        35.0           6.7  1959   
1668          Return from Witch Mountain        38.0           5.6  1978   

          id       est  
522      280  3.143963  
1011     218  3.120799  
8401   54138  3.051255  
974      679  2.931350  
8658  127585  2.905935  
1376     597  2.804721  
922     2756  2.789973  
2014   16306  2.786729  
1621   18887  2.768886  
1668   14822  2.712375

In [ ]:
hybrid_recommender(500, 'Avatar')

title  vote_count  vote_average  year  \
1011                      The Terminator      4208.0           7.4  1984   
1621  Darby O'Gill and the Little People        35.0           6.7  1959   
7088           Star Wars: The Clone Wars       434.0           5.8  2008   
8401             Star Trek Into Darkness      4479.0           7.4  2013   
8658          X-Men: Days of Future Past      6155.0           7.5  2014   
1376                             Titanic      7770.0           7.5  1997   
974                               Aliens      3282.0           7.7  1986   
2014                    Fantastic Planet       140.0           7.6  1973   
3060     Sinbad and the Eye of the Tiger        39.0           6.3  1977   
8724                   Jupiter Ascending      2816.0           5.2  2015   

          id       est  
1011     218  3.380782  
1621   18887  3.269783  
7088   12180  3.232659  
8401   54138  3.202637  
8658  127585  3.188384  
1376     597  3.147407  
974      679  3.122344  
2014   16306  3.108309  
3060   11940  3.011684  
8724   76757  2.992394

- **Simple Recommender:** This system used overall TMDB Vote Count and Vote Averages to build Top Movies Charts, in general and for a specific genre. The IMDB Weighted Rating System was used to calculate ratings on which the sorting was finally performed.

- **Content Based Recommender:** We built two content based engines; one that took movie overview and taglines as input and the other which took metadata such as cast, crew, genre and keywords to come up with predictions. We also deviced a simple filter to give greater preference to movies with more votes and higher ratings.

- **Collaborative Filtering:** We used the powerful Surprise Library to build a collaborative filter based on single value decomposition. The RMSE obtained was less than 1 and the engine gave estimated ratings for a given user and movie.

- **Hybrid Engine:** We brought together ideas from content and collaborative filterting to build an engine that gave movie suggestions to a particular user based on the estimated ratings that it had internally calculated for that user.